text/src/fs2/data/text/CharLikeChunks.scala

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2019 Lucas Satabin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package fs2
+package data
+package text
+
+import scala.annotation.implicitNotFound
+
+import java.nio.charset.Charset
+
+/** A typeclass witnessing that a stream of type `In` has chunks
+  * that can be iterated over to get characters.
+  */
+@implicitNotFound("Cannot prove that stream of type ${In} has chunks that can be iterated over to get characters")
+trait CharLikeChunks[F[_], In] {
+
+  /** The context used to pull chunks and iterate over characters.
+    * Implementations might decide to have immutable or mutable contexts
+    * depending on performance constraints that are desired.
+    */
+  type Context
+
+  /** Creates a context out of a stream. */
+  def create(s: Stream[F, In]): Context
```

Comment:
Maybe it would be better to suspend the creation in an effect? The current way works because every implementation does a `Stream.suspend` and it's all internal, but generally speaking, it shouldn't be the caller's responsibility to take care of that.

Reply:
I don't want to force the `Sync` on `F`.

Reply:
And even if it is an `F` there is no way you can enforce `F.pure` won't be used in this. I prefer doing it for sure at call site. Because here, this really is necessary because of how the call site uses it, so I prefer keeping it that way.

Reply:
Well, as always in Scala you'll find a way to shoot yourself in the foot if you want to – types can only give guidance. But I see your point about the `Sync` constraint, so maybe leave as-is and just add a warning in the Scaladoc about this call contract?

Reply:
Yes, I am writing contribution guidelines currently, including what to be aware of when adding new parsers. This will definitely be part of this as well.